flink-yarn/src/main/java/org/apache/flink/yarn/YarnResourceManager.java

```diff
@@ -0,0 +1,551 @@
+	final static String ENV_FLINK_NODE_ID = "_FLINK_NODE_ID";
+
+	/** Default heartbeat interval between this resource manager and the YARN ResourceManager */
+	private final int yarnHeartbeatIntervalMillis;
+
+	private final Configuration flinkConfig;
+
+	private final YarnConfiguration yarnConfig;
+
+	/** Client to communicate with the Resource Manager (YARN's master) */
+	private AMRMClientAsync<AMRMClient.ContainerRequest> resourceManagerClient;
+
+	/** Client to communicate with the Node manager and launch TaskExecutor processes */
+	private NMClient nodeManagerClient;
+
+	/** The number of containers requested, but not yet granted */
+	private int numPendingContainerRequests;
+
+	public YarnResourceManager(
+			Configuration flinkConfig,
+			Map<String, String> env,
+			RpcService rpcService,
+			ResourceManagerConfiguration resourceManagerConfiguration,
+			HighAvailabilityServices highAvailabilityServices,
+			SlotManagerFactory slotManagerFactory,
+			MetricRegistry metricRegistry,
+			JobLeaderIdService jobLeaderIdService,
+			FatalErrorHandler fatalErrorHandler) {
+		super(
+			rpcService,
+			resourceManagerConfiguration,
+			highAvailabilityServices,
+			slotManagerFactory,
+			metricRegistry,
+			jobLeaderIdService,
+			fatalErrorHandler);
+		this.flinkConfig  = flinkConfig;
+		this.yarnConfig = new YarnConfiguration();
+		this.ENV = env;
+		final int yarnHeartbeatIntervalMS = flinkConfig.getInteger(
+				ConfigConstants.YARN_HEARTBEAT_DELAY_SECONDS, DEFAULT_YARN_HEARTBEAT_INTERVAL_MS / 1000) * 1000;
+
+		final long yarnExpiryIntervalMS = yarnConfig.getLong(
+				YarnConfiguration.RM_AM_EXPIRY_INTERVAL_MS,
+				YarnConfiguration.DEFAULT_RM_AM_EXPIRY_INTERVAL_MS);
+
+		if (yarnHeartbeatIntervalMS >= yarnExpiryIntervalMS) {
+			log.warn("The heartbeat interval of the Flink Application master ({}) is greater " +
+					"than YARN's expiry interval ({}). The application is likely to be killed by YARN.",
+					yarnHeartbeatIntervalMS, yarnExpiryIntervalMS);
+		}
+		yarnHeartbeatIntervalMillis = yarnHeartbeatIntervalMS;
+		numPendingContainerRequests = 0;
+	}
+
+	@Override
+	protected void initialize() throws ResourceManagerException {
+		resourceManagerClient = AMRMClientAsync.createAMRMClientAsync(yarnHeartbeatIntervalMillis, this);
+		resourceManagerClient.init(yarnConfig);
+		resourceManagerClient.start();
+		try {
+			//TODO: change akka address to tcp host and port, the getAddress() interface should return a standard tcp address
+			Tuple2<String, Integer> hostPort = parseHostPort(getAddress());
+			resourceManagerClient.registerApplicationMaster(hostPort.f0, hostPort.f1, getAddress());
+		} catch (Exception e) {
+			LOG.info("registerApplicationMaster fail", e);
+		}
+
+		// create the client to communicate with the node managers
+		nodeManagerClient = NMClient.createNMClient();
+		nodeManagerClient.init(yarnConfig);
+		nodeManagerClient.start();
+		nodeManagerClient.cleanupRunningContainersOnStop(true);
```

Comment:
There are some problems if we do so. First the AMRMClientAsync need a CallbackHandler, but Dispatcher and YarnResourceManager need different CallbackHandlers. And I think what we need in Dispatcher to allocate containers for new JobMasters should be YarnClient.